File: android/lib/src/main/java/org/silvermine/downloadmanager/DownloadRecord.kt
package org.silvermine.downloadmanager

import kotlinx.serialization.Required
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Persisted download record, stored in `downloads.json`. Mirrors the Rust
 * `DownloadRecord` in `crates/download-manager/src/models.rs`.
 *
 * Carries no `progress` — that is derived, and lives only on [DownloadItem],
 * the type sent to the frontend.
 */
@Serializable
internal data class DownloadRecord(
   @SerialName("url")
   val url: String,

   @SerialName("path")
   val path: String,

   @Required
   @SerialName("options")
   val options: CreateOptions = CreateOptions(),

   @SerialName("receivedBytes")
   val receivedBytes: Long = 0L,

   /** `null` when the server did not supply a content length. */
   @SerialName("totalBytes")
   val totalBytes: Long? = null,

   @SerialName("status")
   val status: DownloadStatus = DownloadStatus.Idle,
) {
   /**
    * Sets the byte counts. A `null` total means "this response did not report one",
    * not "there is none": a chunked resume cannot know a length an earlier response
    * already established, so a null never erases a total the record holds. Pass a
    * non-null total to widen or correct it.
    */
   fun withBytes(receivedBytes: Long, totalBytes: Long? = null): DownloadRecord =
      copy(receivedBytes = receivedBytes, totalBytes = totalBytes ?: this.totalBytes)

   fun withStatus(newStatus: DownloadStatus): DownloadRecord =
      copy(status = newStatus)

   /**
    * Builds the public payload, computing `progress` from the byte counts.
    *
    * A completed download always reports 100%, even without a content length.
    * In flight without one it reports 0%, since [receivedBytes] is then the
    * only meaningful signal.
    */
   fun toItem(): DownloadItem {
      val progress = when {
         status == DownloadStatus.Completed -> 100.0
         totalBytes != null && totalBytes > 0L ->
            ((receivedBytes.toDouble() / totalBytes.toDouble()) * 100.0).coerceIn(0.0, 100.0)
         else -> 0.0
      }

      return DownloadItem(
         url = url,
         path = path,
         options = options,
         receivedBytes = receivedBytes,
         totalBytes = totalBytes,
         progress = progress,
         status = status,
      )
   }
}

File: ios/DownloadManagerKit/Sources/DownloadManagerKit/DownloadRecord.swift
//
//  DownloadRecord.swift
//  DownloadManagerKit
//

import Foundation

/// Persisted download record, stored in `downloads.json`. Mirrors the Rust
/// `DownloadRecord` in `crates/download-manager/src/models.rs`.
///
/// Carries no `progress` — that is derived, and lives only on [`DownloadItem`],
/// the type sent to the frontend. `resumeDataPath` is an internal URLSession
/// concern that deliberately never leaves this type.
struct DownloadRecord: Identifiable, Codable, Sendable {
   var id: URL { path }

   let url: URL
   let path: URL
   let options: CreateOptions
   private(set) var receivedBytes: UInt64
   private(set) var totalBytes: UInt64?
   private(set) var status: DownloadStatus
   var resumeDataPath: URL?

   init(
      url: URL,
      path: URL,
      options: CreateOptions = CreateOptions(),
      receivedBytes: UInt64 = 0,
      totalBytes: UInt64? = nil,
      status: DownloadStatus = .idle,
      resumeDataPath: URL? = nil
   ) {
      self.url = url
      self.path = path
      self.options = options
      self.receivedBytes = receivedBytes
      self.totalBytes = totalBytes
      self.status = status
      self.resumeDataPath = resumeDataPath
   }

   /// Sets the byte counts. A nil total means "this callback did not report one",
   /// not "there is none": a resume the server answers without a content length
   /// cannot know a total an earlier response already established, so nil never
   /// erases a total the record holds. Pass a value to widen or correct it.
   mutating func setBytes(received: UInt64, total: UInt64? = nil) {
      self.receivedBytes = received
      self.totalBytes = total ?? self.totalBytes
   }

   mutating func setResumeDataPath(_ resumeDataPath: URL?) {
      self.resumeDataPath = resumeDataPath
   }

   mutating func setStatus(_ status: DownloadStatus) {
      self.status = status
   }

   /// Builds the public payload, computing `progress` from the byte counts.
   ///
   /// A completed download always reports 100%, even without a content length.
   /// In flight without one it reports 0%, since `receivedBytes` is then the
   /// only meaningful signal.
   func toItem() -> DownloadItem {
      let progress: Double

      if status == .completed {
         progress = 100.0
      } else if let total = totalBytes, total > 0 {
         progress = min(max((Double(receivedBytes) / Double(total)) * 100.0, 0.0), 100.0)
      } else {
         progress = 0.0
      }

      return DownloadItem(
         url: url,
         path: path,
         options: options,
         receivedBytes: receivedBytes,
         totalBytes: totalBytes,
         progress: progress,
         status: status
      )
   }
}
